### ISSUE TITLE
Output issues

### ISSUE THREAD

**Issue by jonaskello** [whoeverest/wsrun#3]
Thanks for this tool, it is nice to have something lightweight to just run commands!

Now, when I tried it it worked but I got some issues with the output. This is what I am seeing:

```
$ yarn build2
yarn run v1.3.2
$ wsrun build --stages
>>> @promaster/edit-plugin-adapter $ yarn build
>>> @promaster/edit-plugin-core $ yarn build
>>> @promaster/example-plugin $ yarn build
$ tsc -p tsconfig-build.json
$ tsc -p tsconfig-build.json
$ tsc -p tsconfig-build.json
@promaster/exa  |
@promaster/exa  |
@promaster/edi  |
@promaster/edi  |
@promaster/edi  |
@promaster/edi  |
>>> @promaster/plugin-test-runner $ yarn build
$ tsc -p tsconfig-build.json
@promaster/plu  |
@promaster/plu  |
Done in 8.96s.
```

So basically it is running the `build` cmd for 4 packages. My issues are these:

1. Since I have an organization prefix, the output prefix for each package is too short identity which package the output is for
2. There are blank lines outputted that are not there if I run the same cmd in each repo without wsrun.

**Comment by spion** [comment]
1. Hmm. What if we only showed the last letters of the package name instead of the first ones? What about "@p...plugin-adapter" ? Or should we simply make the columns wide enough to fit most packages?

2. The blank lines are there... its just that they're not visible, since they're the last line of the output. Then the prompt gets shown.

**Comment by wh4everest** [comment]
1. I vote for using first N + "..." + last N letters. Making it wider impacts visibility on terminals with small width.

**Comment by jonaskello** [comment]
Just to provide some more feedback. I currently use [npm-run-all](https://www.npmjs.com/package/npm-run-all)to build my packages (`run-s`). So my root package.json has this:

```
  "scripts": {
    "build2": "wsrun build --stages",
    "build": "run-s -l build:core build:adapter build:plugin build:runner",
    "build:core": "tsc -p ./packages/edit-plugin-core/tsconfig-build.json",
    "build:adapter": "tsc -p ./packages/edit-plugin-adapter/tsconfig-build.json",
    "build:plugin": "tsc -p ./packages/example-plugin/tsconfig-build.json",
    "build:runner": "tsc -p ./packages/plugin-test-runner/tsconfig-build.json",
    "bundle:plugin": "cd ./packages/example-plugin && webpack && cp ./bundle/index.js ../plugin-test-runner/assets/plugins/example-plugin/"
  }
```

The output by npm-run-all is formatted like this:

```
$ yarn build
yarn run v1.3.2
$ run-s -l build:core build:adapter build:plugin build:runner
[build:core   ] $ tsc -p ./packages/edit-plugin-core/tsconfig-build.json
[build:adapter] $ tsc -p ./packages/edit-plugin-adapter/tsconfig-build.json
[build:plugin ] $ tsc -p ./packages/example-plugin/tsconfig-build.json
[build:runner ] $ tsc -p ./packages/plugin-test-runner/tsconfig-build.json
Done in 15.64s.
```
It seems it figures out the widest label and make them all that width. Also it does not contain the extra lines that wsrun does. 

Now I try to change to use wsrun, so instead of having all build cmd in the root package.json I moved them into each the package.json in each package. When I run this with the buld2 cmd from above I get this output:

```
$ yarn build2
yarn run v1.3.2
$ wsrun build --stages
>>> @promaster/edit-plugin-adapter $ yarn build
>>> @promaster/edit-plugin-core $ yarn build
>>> @promaster/example-plugin $ yarn build
$ tsc -p tsconfig-build.json
$ tsc -p tsconfig-build.json
$ tsc -p tsconfig-build.json
@promaster/exa  |
@promaster/exa  |
@promaster/edi  |
@promaster/edi  |
>>> @promaster/plugin-test-runner $ yarn build
@promaster/edi  |
@promaster/edi  |
$ tsc -p tsconfig-build.json
@promaster/plu  |
@promaster/plu  |
Done in 8.87s.
```

I prefer the npm-run-all way of formatting the output so anything that brings wsrun's output closer to that would be nice. 

What is really nice about wsrun though is that it enables me to have the build cmd local in each package and runs some builds in parallel according to the package dependencies. 👍

**Comment by spion** [comment]
Well unfortunately, wsrun can't rename the packages in such a way - it doesn't know which parts of the name it can omit. Additionally if we simply implement scaling to maximum length, the lines will be very big.

The fanciest way would be to split the pkgname into delimiters, and find all words with length > 1 that repeat more than N times in the list, then replace them (and the delimiter) with "..". In your case that might result with:

```
@/e..p..adapter
@/e..p..core
@/example-p..
``

**Comment by jonaskello** [comment]
Not sure what you meant by rename the packages, perhaps I was a bit unclear about having the same output as npm-run-all. In that case it uses the cmd name, but for wsrun it would of course use the package name instead, something like this:

```
$ yarn build2
yarn run v1.3.2
$ run-s -l build:core build:adapter build:plugin build:runner
[@promaster/edit-plugin-core   ] $ tsc -p ./packages/edit-plugin-core/tsconfig-build.json
[@promaster/edit-plugin-adapter] $ tsc -p ./packages/edit-plugin-adapter/tsconfig-build.json
[@promaster/example-plugin     ] $ tsc -p ./packages/example-plugin/tsconfig-build.json
[@promaster/plugin-test-runner ] $ tsc -p ./packages/plugin-test-runner/tsconfig-build.json
Done in 15.64s.
```

Agreed, the lines would be long if it is done like this. The only other options I can see is to have some algorithm like you outlines, or some config to set the length etc.

**Comment by spion** [comment]
Sorry, I meant it can't figure out a short-name version that makes sense! :grin: 

I fixed the multiple line issue, btw. Will push that out soon.

**Comment by spion** [comment]
Both things should be improved in v1.2.3 now.

**Comment by jonaskello** [comment]
Thank you, awesome work!

With v1.2.3 my output looks a lot nicer:

```$ yarn build2
yarn run v1.3.2
$ wsrun build --stages
>>> @promaster/edit-plugin-adapter $ yarn build
>>> @promaster/edit-plugin-core $ yarn build
>>> @promaster/example-plugin $ yarn build
$ tsc -p tsconfig-build.json
$ tsc -p tsconfig-build.json
$ tsc -p tsconfig-build.json
>>> @promaster/plugin-test-runner $ yarn build
$ tsc -p tsconfig-build.json
✨  Done in 5.79s.
```

The package lables are gone but I actually think this is better. npm-run-all does not show labels as standard either, it is an opt in flag. I also tried --collect-output but it seems to give the same output. Not sure if that is correct or not?

**Comment by spion** [comment]
Yes thats expected, however if tsc does generate some output it will show up with the labels.

--collect-output means the stdouts/errs of different processes will be shown all at once instead of intermingled

You can always use --no-prefix if you want the prefix to be gone.

**Comment by jonaskello** [comment]
Aha, I see. So when the CLI command is started that will not get prefixed but if it outputs something it will be prefixed? 

Would it be possible, or/or make sense to prefix all lines?

To show what I mean I put a compile error in each package to get some output:

Output with v1.2.3 using --collect-output

```
$ yarn build2
yarn run v1.3.2
$ wsrun build --stages --collect-ouput
>>> @promaster/edit-plugin-adapter $ yarn build
>>> @promaster/edit-plugin-core $ yarn build
>>> @promaster/example-plugin $ yarn build
$ tsc -p tsconfig-build.json
$ tsc -p tsconfig-build.json
$ tsc -p tsconfig-build.json
@..example-plug | src/index.tsx(2,23): error TS2307: Cannot find module '@promaster/edit-plugin-core2'.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
`yarn build` failed with exit code 1
@..edit-p..core | src/plugin-types/area-plugin.ts(1,28): error TS2307: Cannot find module './plugin-base2'.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
`yarn build` failed with exit code 1
@..e..p..adapte | src/plugin-adapter.tsx(19,65): error TS1005: ',' expected.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
`yarn build` failed with exit code 1
>>> @promaster/plugin-test-runner $ yarn build
$ tsc -p tsconfig-build.json
@..p..test-runn | src/app.tsx(7,4): error TS1005: ',' expected.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
`yarn build` failed with exit code 1
✨  Done in 5.83s.
```

Proposed output that makes the most sense to me:

```
$ yarn build2
yarn run v1.3.2
$ wsrun build --stages --collect-ouput
@..e..p..adapte | $ yarn build
@..e..p..adapte | $ tsc -p tsconfig-build.json
@..e..p..adapte | src/plugin-adapter.tsx(19,65): error TS1005: ',' expected.
@..e..p..adapte | error Command failed with exit code 2.
@..e..p..adapte | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
@..e..p..adapte | `yarn build` failed with exit code 1
@..edit-p..core | $ yarn build
@..edit-p..core | $ tsc -p tsconfig-build.json
@..edit-p..core | src/plugin-types/area-plugin.ts(1,28): error TS2307: Cannot find module './plugin-base2'.
@..edit-p..core | error Command failed with exit code 2.
@..edit-p..core | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
@..edit-p..core | `yarn build` failed with exit code 1
@..example-plug | $ yarn build
@..example-plug | $ tsc -p tsconfig-build.json
@..example-plug | src/index.tsx(2,23): error TS2307: Cannot find module '@promaster/edit-plugin-core2'.
@..example-plug | error Command failed with exit code 2.
@..example-plug | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
@..example-plug | `yarn build` failed with exit code 1
@..p..test-runn | $ yarn build
@..p..test-runn | $ tsc -p tsconfig-build.json
@..p..test-runn | src/app.tsx(7,4): error TS1005: ',' expected.
@..p..test-runn | error Command failed with exit code 2.
@..p..test-runn | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
@..p..test-runn | `yarn build` failed with exit code 1
✨  Done in 5.83s.
```

Sorry for nitpicking on the output, and thanks for taking the time to answer my comments :-).

**Comment by spion** [comment]
Unfortunately yarn adds a control that clears the line when writing to output, and I don't know how to remove it.

**Comment by jonaskello** [comment]
Here is the corresponding output from npm-run-all with the compile errors:

```
$ yarn build
yarn run v1.3.2
$ run-p -l build:core build:adapter build:plugin build:runner
[build:adapter] $ tsc -p ./packages/edit-plugin-adapter/tsconfig-build.json
[build:core   ] $ tsc -p ./packages/edit-plugin-core/tsconfig-build.json
[build:plugin ] $ tsc -p ./packages/example-plugin/tsconfig-build.json
[build:runner ] $ tsc -p ./packages/plugin-test-runner/tsconfig-build.json
[build:plugin ] packages/example-plugin/src/index.tsx(2,23): error TS2307: Cannot find module '@promaster/edit-plugin-core2'.
[build:plugin ] error Command failed with exit code 2.
[build:plugin ] info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
[build:core   ] packages/edit-plugin-core/src/plugin-types/area-plugin.ts(1,28): error TS2307: Cannot find module './plugin-base2'.
[build:core   ] error Command failed with exit code 2.
[build:core   ] info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
[build:runner ] packages/plugin-test-runner/src/app.tsx(7,4): error TS1005: ',' expected.
[build:runner ] error Command failed with exit code 2.
[build:runner ] info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
[build:adapter] error Command failed with signal "SIGTERM".
[build:adapter] info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: "build:plugin" exited with 1.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
I was thinking maybe there are some ideas inside the source code for npm-run-all how to make the yarn output prefixed. It seems to be able to prefix all yarn output except for the last one output.

**Comment by spion** [comment]
Pull requests welcome.

**Comment by jonaskello** [comment]
I moved my suggestion to #4. If I can find the time I might have a look at making a PR.
